tests/bsim/bluetooth/mesh/src/test_proxy_sol.c

```diff
@@ -0,0 +1,304 @@
+{
+	static uint32_t fixed_seq_n = 0x123456;
+
+	net_buf_simple_add_u8(pdu, sub->keys[SUBNET_KEY_TX_IDX(sub)].msg.nid);
+	net_buf_simple_add_u8(pdu, 0x80);
+	net_buf_simple_add_le24(pdu, sys_cpu_to_be24(fixed_seq_n));
+	net_buf_simple_add_le16(pdu, sys_cpu_to_be16(bt_mesh_primary_addr()));
+	net_buf_simple_add_le16(pdu, 0x0000);
+
+	ASSERT_OK(bt_mesh_net_encrypt(&sub->keys[SUBNET_KEY_TX_IDX(sub)].msg.enc, pdu, 0,
+				      BT_MESH_NONCE_SOLICITATION));
+
+	ASSERT_OK(bt_mesh_net_obfuscate(pdu->data, 0,
+					&sub->keys[SUBNET_KEY_TX_IDX(sub)].msg.privacy));
+
+	net_buf_simple_push_u8(pdu, 0);
+	net_buf_simple_push_le16(pdu, BT_UUID_MESH_PROXY_SOLICITATION_VAL);
+}
+
+static int sol_fixed_pdu_send(void)
+{
+	NET_BUF_SIMPLE_DEFINE(pdu, 20);
+	net_buf_simple_init(&pdu, 3);
+
+	struct bt_mesh_subnet *sub = bt_mesh_subnet_find(NULL, NULL);
+
+	uint16_t adv_int = BT_MESH_TRANSMIT_INT(CONFIG_BT_MESH_SOL_ADV_XMIT);
+
+	sol_fixed_pdu_create(sub, &pdu);
+
+	struct bt_data ad[] = {
+		BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+		BT_DATA_BYTES(BT_DATA_UUID16_ALL,
+			      BT_UUID_16_ENCODE(BT_UUID_MESH_PROXY_SOLICITATION_VAL)),
+		BT_DATA(BT_DATA_SVC_DATA16, pdu.data, pdu.size),
+	};
+
+	return bt_mesh_adv_bt_data_send(CONFIG_BT_MESH_SOL_ADV_XMIT, adv_int, ad, 3);
+}
+
+static void test_tester_beacon_rcvd(void)
+{
+	tester_configure();
+
+	/* Check that no beacons are currently being picked up by the scanner */
+	ASSERT_EQUAL(-EAGAIN, k_sem_take(&beacon_sem, K_SECONDS(SEM_TIMEOUT)));
+
+	ASSERT_OK(bt_mesh_proxy_solicit(0));
+
+	ASSERT_OK_MSG(k_sem_take(&beacon_sem, K_SECONDS(SEM_TIMEOUT)),
+		      "No beacon (proxy advs) reveiced.");
+
+	PASS();
+}
+
+static void test_tester_immediate_replay_attack(void)
+{
+	tester_configure();
+
+	/* Check that no beacons are currently being picked up by the scanner */
+	ASSERT_EQUAL(-EAGAIN, k_sem_take(&beacon_sem, K_SECONDS(SEM_TIMEOUT)));
+
+	/* Send initial solicitation PDU with fixed sequence number */
+	ASSERT_OK(sol_fixed_pdu_send());
+
+	ASSERT_OK_MSG(k_sem_take(&beacon_sem, K_SECONDS(SEM_TIMEOUT)),
+		      "No beacon (proxy advs) reveiced.");
+
+	/* Wait for iut proxy advs to time out */
+	k_sleep(K_MSEC(200));
+	ASSERT_EQUAL(-EAGAIN, k_sem_take(&beacon_sem, K_SECONDS(SEM_TIMEOUT)));
+
+	k_sem_reset(&beacon_sem);
+
+	/* Replay attack */
+	ASSERT_OK(sol_fixed_pdu_send());
+
+	ASSERT_EQUAL(-EAGAIN, k_sem_take(&beacon_sem, K_SECONDS(SEM_TIMEOUT)));
+
+	PASS();
+}
+
+static void test_tester_power_replay_attack(void)
+{
+	tester_configure();
+
+	/* Check that no beacons are currently being picked up by the scanner */
+	ASSERT_EQUAL(-EAGAIN, k_sem_take(&beacon_sem, K_SECONDS(SEM_TIMEOUT)));
+
+	/* Replay attack */
+	ASSERT_OK(sol_fixed_pdu_send());
+
+	ASSERT_EQUAL(-EAGAIN, k_sem_take(&beacon_sem, K_SECONDS(SEM_TIMEOUT)));
+
+	/* Send a sol PDU with new sequence number */
+	ASSERT_OK(bt_mesh_proxy_solicit(0));
+
+	ASSERT_OK_MSG(k_sem_take(&beacon_sem, K_SECONDS(SEM_TIMEOUT)),
```

Comment:
How does this work if the internal sseq (https://github.com/zephyrproject-rtos/zephyr/blob/6df6935b674c6a253e3e1ed2628f42821e3b90e7/subsys/bluetooth/mesh/solicitation.c#L40C17-L40C25) is definitely not greater than the fixed one (`0x123456`). Does it bypasses this check because the initial sseq value is 0:
https://github.com/zephyrproject-rtos/zephyr/blob/6df6935b674c6a253e3e1ed2628f42821e3b90e7/subsys/bluetooth/mesh/solicitation.c#L68

Reply:
Good question! I will investigate.

Reply:
It appears to be functioning as designed imo.

If sseq is zero, the condition evaluates to false, enabling the initial setup of sseq to zero for a new sender. This allows the function to proceed with updating the stored sequence number in the SRPL for the first message from a new sender.

Reply:
Perhaps the check should be `entry->sseq > sseq || (entry->sseq == sseq && sseq != 0)`?

Reply:
Sorry, no, I'm wrong. This will allow to receive the same message with SSEQ zero. Maybe `entry->sseq >= sseq`?

Reply:
Can you give an example of a case where you think that the sequence number/message is not handled correctly with the current implementation? Also, if changes are needed here, should it be done in a separate PR?

Reply:
The example is this test. The test sends a message with some SSEQ value. Then it sends another message with the SSEQ lower that the last stored. By the spec, if a node has a stored SRPL entry for this source address and the stored SSEQ is lower than the one from the message, the message shall be discarded (section 6.9.2.1):
![image](https://github.com/zephyrproject-rtos/zephyr/assets/677018/0b20be2c-8701-42ad-a9f6-d035322b8473)
But our implementation always accepts a message with SSEQ zero.

Reply:
So, in the test, what I would change is to set the SSEQ in the manually generated solicitation message to for example 2, then after sending that message, the tester can use `bt_mesh_proxy_solicit` function to check that SSEQs lower than 2 are ignored. It, of course, needs to call `bt_mesh_proxy_solicit` 3 times because the internal SSEQ value is still zero after sending the manually generated solicitation message.

Reply:
Right. How about `(entry->sseq >= sseq || (entry->sseq == 0 && sseq != 0))`? It will handle the first sequence number from a new sender when it can potentially be zero, while still enforcing the requirement of increasing sequence numbers for subsequent messages.

Reply:
If the entry doesn't exist `srpl_find_by_addr` will return NULL. If the entry exists then the stored SSEQ can be any value including zero. This means that SSEQ zero  can't be accepted in any case. Then `entry->sseq >= sseq` should be sufficient.

Reply:
@alxelax, do you agree with this solution?

Reply:
Yep, @PavelVPV's suggestion is correct.